### PR TITLE
Guard viper configuration with RWLock

### PR DESCRIPTION
**Pull request by mdrohmann**
https://www.pivotaltracker.com/story/show/176721598

### CODE REVIEW

**Comment by mdrohmann** [comment]
It is hard to tell, but I am fairly sure that if this fixed the problem.  I could re-produce it once more without these changes (took a lot of tries), but now it does not seem to happen anymore...

File: internal/analytics/deferred.go [hunk 1]
```diff
@@ -122,6 +122,9 @@ func sendDeferred(cfg Configurable, sender func(string, string, string, map[stri
 	if err != nil {
 		return errs.Wrap(err, "Could not load events on send")
 	}
+	if len(deferred) == 0 {
```
Comment (daved) [inline internal/analytics/deferred.go:125]: Suggestion: It might be worth, instead, using `if len(deferred) > 0` and placing it around the cfg.Save below so that the "deferrerFilePath" is cleaned up.

**Review by daved — DISMISSED** [review]

**Review by daved — APPROVED** [review]

File: internal/analytics/deferred.go [hunk 2]
```diff
@@ -122,17 +122,19 @@ func sendDeferred(cfg Configurable, sender func(string, string, string, map[stri
 	if err != nil {
 		return errs.Wrap(err, "Could not load events on send")
 	}
-	for n, event := range deferred {
-		if err := sender(event.Category, event.Action, event.Label, event.Dimensions); err != nil {
-			return errs.Wrap(err, "Could not send deferred event")
+	if len(deferred) > 0 {
```
Comment (Naatan) [inline internal/analytics/deferred.go:125]: Nit: This is an unnecessary change.

**Review by Naatan — APPROVED** [review]

File: internal/analytics/deferred.go (hunk 2 above)
Comment (mdrohmann) [reply internal/analytics/deferred.go:125]: It skips the save of the configuration file, which I believe should be avoided if nothing changed.  Of course, it doesn't affect this bug, but as we have seen that viper also has issues with concurrent writes between processes...
